### ISSUE TITLE
PG does not use binpath

### ISSUE THREAD

**Issue by sampollard** [ProofGeneral/PG#450]
Forgive me if this is a duplicate, it appears there is a TODO [here](https://github.com/ProofGeneral/PG/blob/15ccaec24ce935de366cae08b906c130379758ce/coq/coq-system.el#L619)

The situation is I'm using a tactic which calls an external binary, in this case [gappa](http://gappa.gforge.inria.fr/). My `_CoqProject` looks like this (with `$HOME` expanded)
```
-Q $HOME/.opam/default/lib/coq/user-contrib/Gappa Gappa
-bindir $HOME/.local/bin
```
When compiling a simple program via `coqc`, this works. However, stepping through the program interactively I get the following error:

```
Error:
In nested Ltac calls to "gappa", "abstract (gappa_prepare; gappa_internal)",
"gappa_internal" and "<gappatac::gappatac_gappa_internal@0>", last call failed.
execution of Gappa failed:
sh: gappa: command not found
```

Adding this to my `.emacs` didn't help either:
```
(setenv "COQBIN" (concat (getenv "HOME") "/.local/bin"))
```

The [Coq manual](https://coq.inria.fr/refman/practical-tools/coq-commands.html#by-environment-variables) has this to say:

> Some Coq commands call other Coq commands. In this case, they look for the commands in directory specified by $COQBIN. If this variable is not set, they look for the commands in the executable path.

I assume this means your `$PATH` environment variable, so a workaround is to add the path directly using 
```
(setenv "PATH" (concat (getenv "PATH") ":" (getenv "HOME") "/.local/bin"))
```
but this is not a project-independent solution.

**Comment by hendriktews** [comment]
Hi,

> Adding these lines to my .emacs didn't help either:
>
> (add-to-list 'exec-path (concat (getenv "HOME") "/.local/bin"))

gappa is called from coq, therefore setting the emacs internal
exec-path cannot help.

> (setenv "COQBIN" (concat (getenv "HOME") "/.local/bin"))

If the coq documentation you cited is right, then this should
work. But you need to execute this before coq is started from
Proof General, i.e. before you open any coq file, to be on the
save side.

> I assume this means your $PATH environment variable.

Yes, this means PATH.

> A workaround is to add the path directly using (setenv "PATH"
> (concat (getenv "PATH") ":" (getenv "HOME") "/.local/bin")) but
> this is not a project-independent solution.

I agree that Proof General should probably process the setting in
_CoqProject. @Pierre, what is your opinion?

Best,

Hendrik

**Comment by cpitclaudel** [comment]
Seems reasonable to me. The code that processes CoqProject should be reasonably easy to extend. But I think seeing path without using CoqProject is reasonable too.  Can you clarify what you mean when you say it's not project-independent?

Also, this bindir option sounds pretty scary to me, security-wise.

**Comment by Matafou** [comment]
PG already reads Coqproject file. If this is an officially supported coqtop option we should support it too. bit is it? If it is I am surprised it does not work already.

**Comment by sampollard** [comment]
Thanks everyone for your responses.

>  Can you clarify what you mean when you say it's not project-independent?
I just mean that I wouldn't have to modify my `.emacs` for each project by modifying `BINPATH`, instead it could be read from `_CoqProject`.

However, having something implicitly change the path could be a security concern, and one line per project is not very inconvenient.

> If the coq documentation you cited is right, then this should work. But you need to execute this before coq is started from Proof General, i.e. before you open any coq file, to be on the safe side.

Putting the `(setenv "COQBIN" (concat (getenv "HOME") "/.local/bin"))` line early in my `.emacs` doesn't work, is that what you mean?

To wrap up, I can get what I need by setting my `PATH` in my `.emacs` so I am fine with closing this issue if need be, I just wanted to document my solution.

**Comment by Matafou** [comment]
@sampollard I think you pointed a tral lack of feature. However _CoqProject files are a coq feature that we sync on. It would be unpolite to add features in it without asking the coq team. Moreover this file is supposed to be versioned with your coq project so any absolute file name should be banned.
But this needs a better workaround anyway.

**Comment by hendriktews** [comment]
Hi Sam,

I would be interested in why setting COQBIN from inside emacs
does not work. If you are interested in this as well, then we can
continue to discuss this on this issue. If you feel you have
spent too much time on this already, then silently stopping to
answer at any point of your choice is totally fine with me...

> Putting the (setenv "COQBIN" (concat (getenv "HOME")
> "/.local/bin")) line early in my .emacs doesn't work, is that
> what you mean?

From this I am tempted to conclude that either this setenv form
does not get executed in your emacs or that COQBIN does not work
as you described in your first message.

In a fresh shell, where COQBIN has not been set, coqc on the file
containing a call to gappa does fail with the same error you
reported in your first message, right?

Can you confirm that, in the same shell, after setting COQBIN,
coqc on the same file succeeds?

Best regards,

Hendrik

**Comment by sampollard** [comment]
Hendrik,

I'm happy to look into this more, Proof general has been really useful to me and is a great tool.

Here's what happens on a fresh terminal, both with an unset COQBIN and a PATH that doesn't contain `gappa` and I call coqc:
```
In nested Ltac calls to "gappa", "abstract (gappa_prepare; gappa_internal)",
"gappa_internal" and "<gappatac::gappatac_gappa_internal@0>", last call
failed.
execution of Gappa failed:
sh: gappa: command not found
```

Now, if I set COQBIN but not PATH I _still_ get that error (likewise if I set `-bindir` in _CoqProject).

This leads me to believe it is not a proof general issue but perhaps my own misunderstanding of how coq works with external binaries. I don't know how Coq does this under the hood, but the documentation (and name) suggests COQBIN is only for other coq files and not something like `gappa` (which, I believe, is an OCaml binary which does some sort tactics-style proof search).
